src/blas/backends/cublas/cublas_helper.hpp

```diff
@@ -176,6 +176,12 @@ class cuda_error : virtual public std::runtime_error {
         throw cublas_error(std::string(#name) + std::string(" : "), err); \
     }
 
+#define CUBLAS_ERROR_FUNC_T(name, func, err, ...)                        \
```

Comment:
I like this idea. Is it possible to completely replace the `CUBLAS_ERROR_FUNC` macro?

Reply:
It depends, some cublas functions are invoked directly without passing from a template parameter representing a function pointer, e.g.,
https://github.com/oneapi-src/oneMKL/blob/f43e4e5943d0ef3a85ff3331d567eadb09c8fe71/src/blas/backends/cublas/cublas_level1.cpp#L381

In that case, maintaining the  ``CUBLAS_ERROR_FUNC`` macro will avoid repeating the cublas function name as a string, i.e.,
```
CUBLAS_ERROR_FUNC_T("cublasSdot", cublasSdot, ....
```

Reply:
I vote for keeping a single macro if possible. Using function name as a string for a few cases (3 if I counted correctly) looks cleaner than having two macros.

Reply:
In reality, they are little bit more.

```
src/blas/backends/cublas/cublas_level1.cpp:384:            CUBLAS_ERROR_FUNC(cublasSdot, err, handle, n, x_, incx, y_, incy, res_);
src/blas/backends/cublas/cublas_level1.cpp:960:            CUBLAS_ERROR_FUNC(cublasSdot, err, handle, n, x_, incx, y_, incy, res_);
src/blas/backends/cublas/cublas_scope_handle.cpp:74:        CUBLAS_ERROR_FUNC(cublasDestroy, err1, handle);
src/blas/backends/cublas/cublas_scope_handle.cpp:99:                CUBLAS_ERROR_FUNC(cublasGetStream, err, handle, &currentStreamId);
src/blas/backends/cublas/cublas_scope_handle.cpp:101:                    CUBLAS_ERROR_FUNC(cublasSetStream, err, handle, streamId);
src/blas/backends/cublas/cublas_scope_handle.cpp:113:    CUBLAS_ERROR_FUNC(cublasCreate, err, &handle);
src/blas/backends/cublas/cublas_scope_handle.cpp:114:    CUBLAS_ERROR_FUNC(cublasSetStream, err, handle, streamId);
src/blas/backends/cublas/cublas_handle.hpp:39:                    CUBLAS_ERROR_FUNC(cublasDestroy, err, handle);
src/blas/backends/cublas/cublas_level3.cpp:98:            CUBLAS_ERROR_FUNC(cublasGemmEx, err, handle, get_cublas_operation(transa),
src/blas/backends/cublas/cublas_level3.cpp:494:            CUBLAS_ERROR_FUNC(cublasGemmEx, err, handle, get_cublas_operation(transa),
src/blas/backends/cublas/cublas_scope_handle_hipsycl.cpp:47:                CUBLAS_ERROR_FUNC(cublasGetStream, err, handle, &currentStreamId);
src/blas/backends/cublas/cublas_scope_handle_hipsycl.cpp:49:                    CUBLAS_ERROR_FUNC(cublasSetStream, err, handle, streamId);
src/blas/backends/cublas/cublas_scope_handle_hipsycl.cpp:60:    CUBLAS_ERROR_FUNC(cublasCreate, err, &handle);
src/blas/backends/cublas/cublas_scope_handle_hipsycl.cpp:61:    CUBLAS_ERROR_FUNC(cublasSetStream, err, handle, streamId)
```

I think that having two different macros could produce a nicer code in ```cublas_scope_handle*```.

In any case, I'm open to modify it.

Reply:
@pgorlani I did not realize the other calls you listed. I am fine with either way as well. I guess we can keep it as it is.

Reply:
Yes, thanks @mmeterel!